feat: add optimal pipeline setting post (#385)
hendriknielaender/double-trouble: data/blog/optimal-pipeline-setup.md

---
publishDate: "Jan 21 2024"
title: "How to setup efficient Pipeline Triggers"
description: "How we tune our CI/CD pipelines for optimal workflows and minimal waiting times"

> image: "~/assets/images/thumbnails/optimal_pipeline_setup_dalle.png"

Author: need a better picture

Reviewer: ![v1](https://github.com/hendriknielaender/double-trouble/assets/51416554/6668d992-227d-476b-ba40-739a103896d9)
![7](https://github.com/hendriknielaender/double-trouble/assets/51416554/d02255b2-a10f-4528-98bb-481cee265f42)
![5](https://github.com/hendriknielaender/double-trouble/assets/51416554/eb3f61c8-0cb6-43c0-9935-9799af4acdc3)
![4](https://github.com/hendriknielaender/double-trouble/assets/51416554/0c6c2b9d-1e82-4a5d-88d9-fcb0eccdf588)
![3](https://github.com/hendriknielaender/double-trouble/assets/51416554/802343fd-4d01-408f-8029-0b52e33872a5)
![2](https://github.com/hendriknielaender/double-trouble/assets/51416554/92eac2fb-2222-4218-974b-9986e9179bc0)

Author: Got some more suggestions..
<img width="549" alt="image" src="https://github.com/hendriknielaender/double-trouble/assets/11775168/b9905f1a-4cab-43cb-b35b-046117cafeb8">
<img width="549" alt="image" src="https://github.com/hendriknielaender/double-trouble/assets/11775168/0b034c57-cbb7-4354-aae7-98e10965b02a">
![Uploading image.png…]()

Reviewer: ![9](https://github.com/hendriknielaender/double-trouble/assets/51416554/fa111080-ce08-4ce7-b606-130cc4aa00d7)
![15](https://github.com/hendriknielaender/double-trouble/assets/51416554/c14879ce-e830-499c-8be3-97226184fd77)
![14](https://github.com/hendriknielaender/double-trouble/assets/51416554/d8f5fd2a-293f-41c7-818a-dba02169981e)
![13](https://github.com/hendriknielaender/double-trouble/assets/51416554/134f614c-53dd-4947-af9d-48bca7ccd848)
![12](https://github.com/hendriknielaender/double-trouble/assets/51416554/5ee37886-c22c-4cfe-b696-a36e5dda1818)
![11](https://github.com/hendriknielaender/double-trouble/assets/51416554/e1f7a1cf-b84a-4476-a6f9-4b141b010da4)
![10](https://github.com/hendriknielaender/double-trouble/assets/51416554/dda5c957-7ae2-4ef6-ad73-fc345b0993c5)

Author: Thanks for adding these extra picture, I think it was worth it 💪🏻

---
publishDate: "Jan 28 2024"

> title: "How to setup efficient Pipeline Triggers"

Reviewer: SEO:
```suggestion
title: ""Efficient CI/CD Pipeline Triggers: A Step-by-Step Guide for GitLab"
```

> description: "How we tune our CI/CD pipelines for optimal workflows and minimal waiting times"

Reviewer: SEO: Provide more descriptive and engaging content to entice clicks.
```suggestion
description: "Discover strategies to fine-tune your CI/CD pipelines in GitLab for maximum efficiency and reduced wait times, complete with practical examples and tips."
```

image: "~/assets/images/thumbnails/optimal_pipeline_dalle.jpg"
imageCreditUrl: https://labs.openai.com

> tags: [cloud, infrastructure, IaC, gitlab]

Reviewer: ```suggestion
tags: [cloud, infrastructure, IaC, GitLab, CI/CD, DevOps, pipeline, automation]
```

---

At work we have spend a bunch of time together, looking the pipeline of a massive monorepo we were
handling. Through countless hours of doing this, we've arrived at a few "good defaults", or useful
techniques, which we also found useful in other repos. In this post we will share these general
techniques with concrete examples for gitlab, but similar concepts could also be applied to other
CI/CD platforms.

## Caching & Artifacts

Many pipeline systems have the concept of caching and artifacts, and so does
[gitlab](https://docs.gitlab.com/ee/ci/caching/). Our recommendation is to use caches for install
dependencies and artifacts for build results, with fallback caches to master in case the

> dependencies didn't change. This allows pull requests to skip the install job in case no dependency

Reviewer: ```suggestion
dependencies didn't change. "This setup enables pull requests to bypass the installation job when there are no changes in dependencies."
```

> has changed. Similarly the artifact has to be build only once.

Reviewer: ```suggestion
has changed. Similarly, the artifact has to be build only once.
```

```yaml
example-job:
  script:
    - yarn install --frozen-lockfile --check-files
  cache:
    paths:
      - node_modules/
```

```yaml
variables:
  FF_USE_FASTZIP: "true"
  # These can be specified per job or per pipeline (slowest, slow, default, fast, and fastest)
  ARTIFACT_COMPRESSION_LEVEL: "fastest"
  CACHE_COMPRESSION_LEVEL: "fastest"
```

```yaml
  needs:
    - job: install
      optional: true
```

## Deployment Job Triggers

The deployment to prod on master generally should be a manual action that can be triggered
immediately. It doesn't have to wait for the dev and test deployment to succeed again first. A
merged MR will anyway have had all necessary safety checks succeeding in the Pull Request

> already. This make it safe to deploy immediately to prod after merge in 99% of the cases. This

Reviewer: ```suggestion
already. This make it safe to deploy immediately to production after merge in 99% of the cases. This
```

the job that my workflows needs will be executed? If any use-case somehow needs a bunch of
unrelated other jobs to be executed first, then this can be optimized. The ideal waiting time is
0.

## Job Dependencies and Stages

Adding more and more stages makes the pipeline easy to understand and satisfactory to look at, but
often it comes at the detriment of speed. Therefor removing and merging stages might yield a
faster pipeline, which is still understandable.

When tests are fast, it might even be advantageous to execute the test together with the build
command, instead of spinning up a whole new docker environment in a subsequent action. In other
cases it might make sense to merge the install job with the build or similar. An optimal pipeline
is one that adapts to changes in the repository, to be always as fast as possible.

## The Build Container

If you use big dependencies during deployment, it can be worth it to bundle them already into the
build container. Maintaining 1 or 2 build containers is usually within reason, while it should of
course not go out of hand.

If all you need is the aws cli, note that there is an official docker images for that by AWS

> [link](https://hub.docker.com/r/amazon/aws-cli). Using a combination of such tool-specific

Reviewer: ```suggestion
[AWS CLI Official Docker Image](https://hub.docker.com/r/amazon/aws-cli). Using a combination of such tool-specific
```
